Update metrics for .NET 8 changes

noahfalk opened the pull request:
- Adds instructions for using IMeterFactory with Dependency Injection (#36749)
- Adds testing examples for MetricCollector (#36749)
- Updates naming guidance to use OpenTelemetry conventions (Fixes #36750)


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/diagnostics/metrics-instrumentation.md](https://github.com/dotnet/docs/blob/ac14f40d6c392ba7fca72ab380b999965e0f95bd/docs/core/diagnostics/metrics-instrumentation.md) | [Creating Metrics](https://review.learn.microsoft.com/en-us/dotnet/core/diagnostics/metrics-instrumentation?branch=pr-en-us-36781) |


<!-- PREVIEW-TABLE-END -->

noahfalk committed:
Update metrics for .NET 8 changes

- Adds instructions for using IMeterFactory with Dependency Injection
- Adds testing examples for MetricCollector
- Updates naming guidance to use OpenTelemetry conventions

noahfalk commented:
@IEvangelist - I'm seeing lint errors but if I understand them correctly they are in other files that this PR isn't touching. Am I supposed to fix those too?

noahfalk commented:
cc @tarekgh @JamesNK 
James, looking at the text I think its probably good that ASP.NET docs have more ASP.NET-specific examples but happy to get your feedback on any of it.

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -72,20 +72,20 @@ Press any key to exit
-name that contains the assembly name and optionally a subcomponent name. If an assembly is adding instrumentation for code in a second, independent assembly, the name
-should be based on the assembly that defines the Meter, not the assembly whose code is being instrumented.
+- The name passed to the <xref:System.Diagnostics.Metrics.Meter> constructor should be unique to distinguish it from other Meters. We recommend
+[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines)
```
```suggestion
[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines),
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -72,20 +72,20 @@ Press any key to exit
-should be based on the assembly that defines the Meter, not the assembly whose code is being instrumented.
+- The name passed to the <xref:System.Diagnostics.Metrics.Meter> constructor should be unique to distinguish it from other Meters. We recommend
+[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines)
+which use dotted hierarchical names. Assembly names or namespace names for code being instrumented are usually a good choice. If an assembly is adding instrumentation
```
```suggestion
which use dotted hierarchical names. Assembly names or namespace names for code being instrumented are usually a good choice. If an assembly adds instrumentation
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -72,20 +72,20 @@ Press any key to exit
-ecosystems have encouraged using '.' or '_' as the separator. Microsoft's suggestion is to use '-' in code and let the metric consumer such as OpenTelemetry or
-Prometheus convert to an alternate separator if needed.
+- .NET doesn't enforce any naming scheme for Instruments, but we recommend following
+[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines) which use dotted hierarchical names
```
```suggestion
[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines), which use dotted hierarchical names
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -72,20 +72,20 @@ Press any key to exit
-Prometheus convert to an alternate separator if needed.
+- .NET doesn't enforce any naming scheme for Instruments, but we recommend following
+[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines) which use dotted hierarchical names
+and an '_' as the separator between multiple words in the same element. Beware that not all metric tools preserve the Meter name as part of the final metric name so it is beneficial
```
```suggestion
and an underscore ('_') as the separator between multiple words in the same element. Not all metric tools preserve the Meter name as part of the final metric name, so it's beneficial
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -103,33 +103,89 @@ to install it:
 
 As expected, you can see that HatCo store is steadily selling 4 hats each second.
 
+## Getting a Meter via Dependency Injection
```
```suggestion
## Get a Meter via dependency injection
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -103,33 +103,89 @@ to install it:
 
+## Getting a Meter via Dependency Injection
+
+In the example above we obtained the Meter by constructing it with `new` and assigning it to a static field. Using statics this way is not a good approach when using Dependency
```
```suggestion
In the previous example, the Meter was obtained by constructing it with `new` and assigning it to a static field. Using statics this way is not a good approach when using dependency
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -103,33 +103,89 @@ to install it:
+## Getting a Meter via Dependency Injection
+
+In the example above we obtained the Meter by constructing it with `new` and assigning it to a static field. Using statics this way is not a good approach when using Dependency
+Injection (DI). In code that uses DI such as ASP.NET Core, or apps with [Generic Host](../extensions/generic-host.md) create the Meter object using
```
```suggestion
injection (DI). In code that uses DI, such as ASP.NET Core or apps with [Generic Host](../extensions/generic-host.md), create the Meter object using
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -103,33 +103,89 @@ to install it:
+
+In the example above we obtained the Meter by constructing it with `new` and assigning it to a static field. Using statics this way is not a good approach when using Dependency
+Injection (DI). In code that uses DI such as ASP.NET Core, or apps with [Generic Host](../extensions/generic-host.md) create the Meter object using
+<xref:System.Diagnostics.Metrics.IMeterFactory>. Starting in .NET 8 hosts will automatically register <xref:System.Diagnostics.Metrics.IMeterFactory> in the service container
```
```suggestion
<xref:System.Diagnostics.Metrics.IMeterFactory>. Starting in .NET 8, hosts will automatically register <xref:System.Diagnostics.Metrics.IMeterFactory> in the service container
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -103,33 +103,89 @@ to install it:
+Injection (DI). In code that uses DI such as ASP.NET Core, or apps with [Generic Host](../extensions/generic-host.md) create the Meter object using
+<xref:System.Diagnostics.Metrics.IMeterFactory>. Starting in .NET 8 hosts will automatically register <xref:System.Diagnostics.Metrics.IMeterFactory> in the service container
+or you can manually register the type in any <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection> by calling <xref:Microsoft.Extensions.DependencyInjection.MetricsServiceExtensions.AddMetrics%2A>.
+The meter factory integrates metrics with DI, keeping Meters in different service collections isolated from each other even if they are using an identical name. This is
```
```suggestion
The meter factory integrates metrics with DI, keeping Meters in different service collections isolated from each other even if they use an identical name. This is
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -103,33 +103,89 @@ to install it:
+<xref:System.Diagnostics.Metrics.IMeterFactory>. Starting in .NET 8 hosts will automatically register <xref:System.Diagnostics.Metrics.IMeterFactory> in the service container
+or you can manually register the type in any <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection> by calling <xref:Microsoft.Extensions.DependencyInjection.MetricsServiceExtensions.AddMetrics%2A>.
+The meter factory integrates metrics with DI, keeping Meters in different service collections isolated from each other even if they are using an identical name. This is
+especially useful for testing so that multiple tests running in parallel will only observe measurements produced from within the same test case.
```
```suggestion
especially useful for testing so that multiple tests running in parallel only observe measurements produced from within the same test case.
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
````diff
@@ -103,33 +103,89 @@ to install it:
+builder.Services.AddSingleton<HatCoMetrics>();
+```
+
+Inject the metrics type and record values where needed. Because the metrics type is registered in DI it can be used with MVC controllers, minimal APIs, or any other type that is created by DI:
````
```suggestion
Inject the metrics type and record values where needed. Because the metrics type is registered in DI, it can be used with MVC controllers, minimal APIs, or any other type that is created by DI:
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -471,3 +524,86 @@ Press p to pause, r to resume, q to quit.
 > [!NOTE]
 > OpenTelemetry refers to tags as 'attributes'. These are two different names for the same functionality.
+
+## Testing custom metrics
```
```suggestion
## Test custom metrics
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -471,3 +524,86 @@ Press p to pause, r to resume, q to quit.
+Its possible to test any custom metrics you add using <xref:Microsoft.Extensions.Telemetry.Testing.Metering.MetricCollector%601>. This type makes it easy to record the measurements
+from specific instruments and assert the values were correct.
+
+### Testing with Dependency Injection
```
```suggestion
### Test with dependency injection
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -471,3 +524,86 @@ Press p to pause, r to resume, q to quit.
+
+### Testing with Dependency Injection
+
+An example test case for code components that use Dependency Injection and IMeterFactory:
```
```suggestion
The following code shows an example test case for code components that use dependency injection and IMeterFactory.
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -471,3 +524,86 @@ Press p to pause, r to resume, q to quit.
+
+Each MetricCollector object records all measurements for one Instrument. If you need to verify measurements from multiple instruments, create one MetricCollector for each one.
+
+### Testing without Dependency Injection
```
```suggestion
### Test without dependency injection
```

tarekgh commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -471,3 +524,86 @@ Press p to pause, r to resume, q to quit.
+    // a host or some other application setup code to do this as well.
+    private static IServiceProvider CreateServiceProvider()
+    {
+        ServiceCollection collection = new ServiceCollection();
```
>collection [](http://example.com/codeflow?start=26&length=10)

nit: will be better calling it `servceCollection`.

gewarren replied:
But don't forget the `i`.

tarekgh replied:
I fixed it :-)

tarekgh reviewed (approved):
LGTM

gewarren commented:
I will try to fix the lint errors separately.

gewarren commented:
> I will try to fix the lint errors separately.

These are fixed now, so they'll go away the next time the PR builds.

noahfalk committed:
Apply suggestions from code review

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

noahfalk committed:
Feedback from PR

JamesNK commented on docs/core/diagnostics/metrics-instrumentation.md:
````diff
@@ -103,33 +103,89 @@ to install it:
+<xref:System.Diagnostics.Metrics.MeterFactoryExtensions.Create%2A>. This example shows adding IMeterFactory to an ASP.NET Core MVC controller type:
+
+```cs
+public class HatStoreController : Controller
+{
+    private readonly Counter<int> _hatsSold;
+
+    public HatStoreController(IMeterFactory meterFactory)
+    {
+        var meter = meterFactory.Create("HatCo.Store");
+        _hatsSold = meter.CreateCounter<int>("hatco.store.hats_sold");
+        // Other methods in this class can now use the _hatsSold counter to increment the count.
+    }
````
I'm not sure we'd want to suggest this as an option at all. A new `Controller` instance is created for every request.

I know Meter and counters are cached, so it doesn't matter too much, but it feels clean to "create" these types once in their own type like is suggested immediately below.

noahfalk replied:
Removed

JamesNK commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -198,18 +254,18 @@ using System.Threading;
-    static Histogram<int> s_orderProcessingTimeMs = s_meter.CreateHistogram<int>("order-processing-time");
+    static Meter s_meter = new Meter("HatCo.Store");
+    static Counter<int> s_hatsSold = s_meter.CreateCounter<int>("hatco.store.hats_sold");
+    static Histogram<int> s_orderProcessingTimeMs = s_meter.CreateHistogram<int>("hatco.store.order_processing_time");
```
Change to seconds? Change type to long? Also, add "s" as unit

JamesNK replied:
I see this in recommendations:

> For timing things, Histogram is usually preferred. Often it's useful to understand the tail of these distributions (90th, 95th, 99th percentile) rather than averages or totals.

Recommend seconds?

noahfalk replied:
I changed it to seconds, but didn't include seconds in the unit because we haven't reached a discussion of units in the doc yet.

JamesNK commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -471,3 +524,86 @@ Press p to pause, r to resume, q to quit.
+        var metrics = new HatCoMetricsWithGlobalMeter();
+        // Be careful specifying scope=null. This binds the collector to a global Meter and tests
+        // that use global state should not be configured to run in parallel.
+        var collector = new MetricCollector<int>(null, "HatCo.Store", "hatco.store.hats_sold");
```
Should `MetricCollector` have a ctor that doesn't have a scope argument?

noahfalk replied:
It does have ctors like (Meter, string) and (Instrument), but it was intentional not to offer a ctor that takes a meter name string binds in global scope by default.

noahfalk committed:
- Removed the example with Controller

- Updated units (where specified) to use OTel conventions for units
- Updated tags to use OTel conventions for naming

noahfalk committed:
- Add advice about seconds for time

noahfalk committed:
Fix link

JamesNK commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -376,12 +376,13 @@ has replaced the generic term "Count" that is visible in previous descriptions.
-- .NET APIs allow any string to be used as the unit, but we recommend using [UCUM](https://ucum.org/), an international standard for unit names. This is also what
-[OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#instrument-units) recommend. The curly
+- .NET APIs allow any string to be used as the unit, but we recommend using [UCUM](https://ucum.org/), an international standard for unit names. The curly
 braces around "{hats}" is part of the UCUM standard indicating that it is a descriptive annotation rather than a unit name with a standardized meaning like seconds or meters.
```
```suggestion
braces around "{hats}" is part of the UCUM standard, indicating that it is a descriptive annotation rather than a unit name with a standardized meaning like seconds or bytes.
```
To avoid confusion about the different meanings of meter.

noahfalk replied:
good call... I didn't even notice that.

noahfalk committed:
Apply suggestions from code review

Co-authored-by: James Newton-King <james@newtonking.com>

noahfalk commented:
OK I switched this to no longer be a draft. I'll wait until end of Friday to merge in case anyone has more comments on the parts I added today. Thanks all!

tommcdon commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -22,7 +22,7 @@ your applications and libraries. In this tutorial, you will add new metrics and
 **Prerequisites**: [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet) or a later version
```
Should we update '.NET Core 3.1' to '.NET 6'?  .NET Core 3.1 is no longer supported.

tommcdon commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -103,33 +103,75 @@ to install it:
 
+## Get a Meter via dependency injection
+
+In the previous example, the Meter was obtained by constructing it with `new` and assigning it to a static field. Using statics this way is not a good approach when using dependency
```
Should we mention that Meter DI was added in .NET 8?

tommcdon replied:
Please disregard - I see that .NET 8 was mentioned earlier.

noahfalk committed:
Update from Net Core 3.1 -> 6 to avoid out-of-support versions
